Reinstate "proxies" support

mgravell opened the pull request:
Exploration into reverting #7705 (the code) and #7510 (the tests) - only the first is reverted yet

mgravell committed:
Revert "removing proxy code (#7705)"

This reverts commit eac199e791fa193084e75cf4b7a04baa8e5f4ab8.

# Conflicts:
#	WebJobs.Script.sln
#	src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs
#	src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
#	src/WebJobs.Script.WebHost/Middleware/FunctionInvocationMiddleware.cs
#	src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj
#	src/WebJobs.Script/Host/HostFunctionMetadataProvider.cs
#	src/WebJobs.Script/Host/ScriptHost.cs
#	src/WebJobs.Script/WebJobs.Script.csproj
#	test/WebJobs.Script.Tests.Integration/ScriptHostEndToEnd/ScriptHostEndToEndTestFixture.cs
#	test/WebJobs.Script.Tests/Security/KeyVaultSecretsRepositoryTests.cs

mgravell committed:
fix package breaks

mgravell committed:
resurrect GetDirectTypes - note: this may be undesirable

mgravell committed:
fixup GetDirectTypes change and removal of BreakingChangeAnalysis

mgravell committed:
without GetDirectTypes, we don't need this suppression

mgravell committed:
remove duplicate test SDK ref

mgravell committed:
reinstate proxies tests

mathewc commented:
@mgravell what's the status of this PR? Are the test failures you're seeing clear to you and are you making progress on them, or do you have questions?

mgravell commented:
@mathewc in progress; I'm talking with Fabio re understanding the test failures; a lot of the failures appear to be related to the hosting of the proxies test target and the related test fixture; I'm struggling with understanding how that *used* to work, so if you have and ideas there: I'm all ears

mathewc commented on test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/ProxyEndToEndTests.cs:
```diff
@@ -0,0 +1,451 @@
+
+                var provider = new HostFunctionMetadataProvider(optionsMonitor, NullLogger<HostFunctionMetadataProvider>.Instance, new TestMetricsLogger());
+
+                var builder = AspNetCore.WebHost.CreateDefaultBuilder()
+                   .UseStartup<Startup>()
+                   .ConfigureServices(services =>
+                   {
+                       services.Replace(new ServiceDescriptor(typeof(IOptions<ScriptApplicationHostOptions>), new OptionsWrapper<ScriptApplicationHostOptions>(HostOptions)));
+                       services.Replace(new ServiceDescriptor(typeof(ISecretManagerProvider), new TestSecretManagerProvider(new TestSecretManager())));
+                       services.Replace(new ServiceDescriptor(typeof(IOptionsMonitor<ScriptApplicationHostOptions>), optionsMonitor));
+                       services.Replace(new ServiceDescriptor(typeof(IFunctionMetadataProvider), provider));
+                   });
+
+                // TODO: https://github.com/Azure/azure-functions-host/issues/4876
+                _testServer = new TestServer(builder);
+                HostOptions.RootServiceProvider = _testServer.Host.Services;
+                var scriptConfig = _testServer.Host.Services.GetService<IOptions<ScriptJobHostOptions>>().Value;
+
+                HttpClient = _testServer.CreateClient();
+                HttpClient.BaseAddress = new Uri("https://localhost/");
+
+                TestHelpers.WaitForWebHost(HttpClient);
```
Syncing your branch, I see that all the tests in this suite are failing because the TestFixture isn't set up properly (the tests have been disabled in this branch for some time, so the code is stale). Replace these highlighted lines with the following:

```csharpTestHost = new TestFunctionHost(HostOptions.ScriptPath, HostOptions.LogPath,
    configureScriptHostServices: services =>
    {
        services.Replace(new ServiceDescriptor(typeof(IOptions<ScriptApplicationHostOptions>), new OptionsWrapper<ScriptApplicationHostOptions>(HostOptions)));
        services.Replace(new ServiceDescriptor(typeof(ISecretManagerProvider), new TestSecretManagerProvider(new TestSecretManager())));
        services.Replace(new ServiceDescriptor(typeof(IOptionsMonitor<ScriptApplicationHostOptions>), optionsMonitor));
        services.Replace(new ServiceDescriptor(typeof(IFunctionMetadataProvider), provider));
    });

HttpClient = TestHost.HttpClient;

TestHelpers.WaitForWebHost(HttpClient);
```

Remove the unused _testServer field, and change Dispose to dispose

mathewc committed:
Reenabling Proxies

mathewc commented on test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/ProxyEndToEndTests.cs:
```diff
@@ -0,0 +1,451 @@
+
+                var provider = new HostFunctionMetadataProvider(optionsMonitor, NullLogger<HostFunctionMetadataProvider>.Instance, new TestMetricsLogger());
+
+                var builder = AspNetCore.WebHost.CreateDefaultBuilder()
```
Pulling your branch, I see these tests don't currently run because the fixture is set up incorrectly. The tests have been disabled for a long time, so likely everything got stale. I've created a branch https://github.com/Azure/azure-functions-host/pull/8602 where I got most of these tests running. Please update your branch accordingly.

Still a handful of proxies tests failing, but at least they run. Need to look at the remaining case by case. Another useful approach would be to enlist in the v3.x branch and for any tests/behaviors in question see how the tests are running in that branch where the feature is enabled to help determine what we need to do to get them running again in v4.x, and what the expected behavior should be.

mgravell committed:
Merge branch 'dev' into reinstate-proxies

mgravell committed:
fixup Validate_HostLogs

mgravell committed:
fixup ProxyFunctionDescriptorProviderTests

mgravell committed:
fixup ValidateProxyFunctionDescriptor

mgravell committed:
update to Gibraltar ("proxies") 2.2.20220829.38

mgravell committed:
Merge branch 'dev' into reinstate-proxies

# Conflicts:
#	src/WebJobs.Script/Host/ScriptHost.cs
#	src/WebJobs.Script/WebJobs.Script.csproj

mgravell committed:
stabilize scripting tests by reverting Microsoft.CodeAnalysis rev from v4 to v3 (required updating Gibraltar)

mgravell committed:
lib rev: revert package updates

mgravell commented:
Failing tests needed work in `Microsoft.Azure.AppService.Proxy.*`, now complete and deployed. The dependencies list needs tweaking; current output:

``` txt
      Changed:
    
      Removed:
    
      Added:
        - Autofac.dll: 4.6.2.0/4.6.2.0
        - Microsoft.AspNetCore.Mvc.Razor.Extensions.dll: 2.1.0.0/2.1.0.18136
        - Microsoft.AspNetCore.Razor.Language.dll: 2.1.0.0/2.1.0.18136
        - Microsoft.Azure.AppService.Proxy.Client.dll: 2.2.2069.1/2.2.2069.1
        - Microsoft.Azure.AppService.Proxy.Common.dll: 2.2.2069.1/2.2.2069.1
        - Microsoft.Azure.AppService.Proxy.Runtime.dll: 2.2.2069.1/2.2.2069.1
        - Microsoft.CodeAnalysis.Razor.dll: 2.1.0.0/2.1.0.18136
        - protobuf-net.dll: 2.3.3.0/2.3.3.0
```

which is the same lib-set that "proxies" used to use

brettsam committed:
updating runtimeassemblies.json; updating test deps.json

mgravell committed:
- gate proxies support via a feature-flag
- enable feature-flag in relevant tests

jviau commented on src/WebJobs.Script/Binding/Http/RawScriptResult.cs:
```diff
@@ -10,8 +10,10 @@
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.AspNetCore.Mvc.Formatters;
 using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Azure.AppService.Proxy.Runtime.Configuration.Policies;
```
Are these usings necessary?

mgravell replied:
fixed in c34f9fe5078d77d670890be613c023113a3862eb

jviau commented on src/WebJobs.Script/Description/Proxies/ProxyFunctionDescriptorProvider.cs:
```diff
@@ -0,0 +1,49 @@
+                return base.TryCreate(functionMetadata);
+            }
+
+            return Task.FromResult<(bool, FunctionDescriptor)>((false, null));
```
Not sure how perf conscious we are or how often this is called, but it might be worth pulling this into a re-usable readonly static. We can avoid extra GC pressure that way.

mgravell replied:
fixed in c34f9fe5078d77d670890be613c023113a3862eb

jviau commented on src/WebJobs.Script/Description/Proxies/ProxyFunctionDescriptorProvider.cs:
```diff
@@ -0,0 +1,49 @@
+        {
+            if (!(functionMetadata is ProxyFunctionMetadata proxyFunctionMetada))
+            {
+                throw new InvalidCastException($"Expected {nameof(functionMetadata)} to be of type {nameof(ProxyFunctionMetadata)}");
```
I think `ArgumentException` is more fitting here. It is more explicit that the caller mis-used this API vs this API made an internal error.

jviau commented on src/WebJobs.Script/Description/Proxies/ProxyFunctionInvoker.cs:
```diff
@@ -0,0 +1,35 @@
+
+            if (requestObj == null)
+            {
+                throw new Exception("Could not find parameter of type HttpRequest while executing a Proxy Request");
```
Is there a more explicit exception type we can throw? Maybe `ArgumentException`?

Also, can do some pattern matching:

``` CSharp
if (parameters?.FirstOrDefault() is not HttpRequest requestObj)
{

}
```

mgravell replied:
pattern matching fixed in c34f9fe5078d77d670890be613c023113a3862eb

jviau commented on src/WebJobs.Script/Description/Workers/WorkerFunctionDescriptorProvider.cs:
```diff
@@ -42,6 +42,12 @@ public override async Task<(bool, FunctionDescriptor)> TryCreate(FunctionMetadat
             }
 
+            // If a function exists exists with a proxy, there is a chance this could get evaluated first before ProxyFunctionDescriptorProvider.
+            if (functionMetadata.IsProxy())
```
As someone new to this codebase I am unsure of all the extensibility models we have. However, it feels a bit odd to me that we need to sprinkle proxy knowledge in other pieces of the code base. Is there anything we can take away from this and hopefully address later?

jviau commented on src/WebJobs.Script/Host/ProxyFunctionProvider.cs:
```diff
@@ -0,0 +1,195 @@
+                return null;
+            }
+
+            string proxiesJson = File.ReadAllText(proxyConfigPath);
```
Given this is called from an async flow (`GetFunctionMetadataAsync`), would it be beneficial to use an async read here?

jviau commented on src/WebJobs.Script/Host/ProxyFunctionProvider.cs:
```diff
@@ -0,0 +1,195 @@
+            return metadata;
+        }
+
+        internal static Collection<ProxyFunctionMetadata> ReadProxyMetadata(string scriptPath, ILogger logger, Dictionary<string, ICollection<string>> functionErrors = null)
```
Does this need to be `internal`?

jviau commented on src/WebJobs.Script/Host/ProxyFunctionProvider.cs:
```diff
@@ -0,0 +1,195 @@
+            {
+                logger.LogInformation("Loading proxies metadata");
+                var metadataCollection = LoadProxyMetadata(proxiesJson, functionErrors, logger);
+                logger.LogInformation($"{metadataCollection.Count} proxies loaded");
```
Consider using structured logging.

```suggestion
                logger.LogInformation("{ProxyCount} proxies loaded", metadataCollection.Count);
```

jviau commented on src/WebJobs.Script/Host/ProxyFunctionProvider.cs:
```diff
@@ -0,0 +1,195 @@
+            string proxiesJson = File.ReadAllText(proxyConfigPath);
+            if (!string.IsNullOrWhiteSpace(proxiesJson))
+            {
+                logger.LogInformation("Loading proxies metadata");
```
nit: this sounds more like a `Trace` or `Debug` than `Information`.

jviau commented on src/WebJobs.Script/Host/ProxyFunctionProvider.cs:
```diff
@@ -0,0 +1,195 @@
+                       .Subscribe(e => HandleProxyFileChange());
+        }
+
+        public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
```
Instead of building this each time, can we just store it as immutable? It looks like `_functionErrors` is written to only once.

mgravell replied:
fixed in c34f9fe5078d77d670890be613c023113a3862eb

jviau commented on src/WebJobs.Script/Host/ProxyFunctionProvider.cs:
```diff
@@ -0,0 +1,195 @@
+            return null;
+        }
+
+        private static Collection<ProxyFunctionMetadata> LoadProxyMetadata(string proxiesJson, Dictionary<string, ICollection<string>> functionErrors, ILogger logger)
```
Can this return `ImmutableArray<T>` directly and use the immutable builder pattern internally?

mgravell replied:
fixed in c34f9fe5078d77d670890be613c023113a3862eb

jviau commented on src/WebJobs.Script/Host/ProxyFunctionProvider.cs:
```diff
@@ -0,0 +1,195 @@
+                client = new ProxyClientExecutor(rawProxyClient);
+            }
+
+            if (client == null)
```
nit: this can be moved up and be conditioned with `if (rawProxyClient is null)`.

mgravell replied:
fixed in c34f9fe5078d77d670890be613c023113a3862eb

jviau commented on src/WebJobs.Script/Host/ProxyFunctionProvider.cs:
```diff
@@ -0,0 +1,195 @@
+                    proxyMetadata.Name = proxyName;
+                    proxies.Add(proxyMetadata);
+                }
+                catch (Exception ex)
```
nit: 
```suggestion
                catch (Exception ex) when (!ex.IsFatal())
```

jviau reviewed (commented):
I understand this is primarily re-introducing previously reverted code. With that in mind, it is up to you if you want to address my comments or leave it as is (as this code was previously already checked in).

liliankasem reviewed (approved):
+1 to Jacob's comments, ideally we can take this opportunity to improve the code we're reintroducing but also understandable if we want to address those in a different PR/issue.

mgravell committed:
Merge branch 'dev' into reinstate-proxies

mgravell commented:
@liliankasem  @jviau  awesome feedback, thanks; and I agree with all of them - but: after discission with @fabiocav , I think we're going to try and merge "as-is", and pick those up as additional work items (pretty small). For context; the nits all relate to the code that we dragged out of the jaws of the shredder.

fabiocav commented:
@jviau and @liliankasem thank you for the review on this!

Had a sync with @mgravell this morning and since there's urgency to get this work in to align with validation and release, and nothing in the feedback is blocking, we'll have this merged and create issues to address the comments. 

@safihamid for awareness

mgravell committed:
fix nits from PR feedback

mgravell commented:
I fixed all the feedback items that have no possible side-effects; merge is blocked on #8741 (dependabot merge broke all builds/CI)

mgravell committed:
Merge branch 'dev' into reinstate-proxies

mgravell committed:
experimental; disabling proxies log checks

mgravell committed:
- move environment-check inside ProxyFunctionProvider to improve testability
- add feature flag to linux/windows E2E tests

mgravell committed:
StandbyModeE2E_Dotnet : perform test with/without proxies - prove that it can be turned off

mgravell committed:
enable proxies in more tests
